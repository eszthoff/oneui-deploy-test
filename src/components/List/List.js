import React from 'react';
import PropTypes from 'prop-types';
import bem from 'bem';
import styles from './List.scss';

const { block } = bem({
    name: 'List',
    classnames: styles,
    propsToMods: ['isDividedList', 'Component']
});

const List = props => {
    const { children, Component, isDividedList, ...rest } = props;
    return (
        <Component {...rest} {...block(props)}>
            {React.Children.map(children, child => React.cloneElement(child, { isDividedList }))}
        </Component>
    );
};

List.displayName = 'List';

List.propTypes = {
    /** List Items */
    children: PropTypes.node,
    /** The HTML or React component to be used to render the list  */
    Component: PropTypes.string,
    /** Adds dividing lines between the list items */
    isDividedList: PropTypes.bool
};

List.defaultProps = {
    children: null,
    Component: 'ul',
    isDividedList: false
};

export default List;
