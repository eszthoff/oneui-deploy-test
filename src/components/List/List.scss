.List {
    &--Component_ul {
        padding: var(--spacing-2x);
        list-style: none;
    }

    &--Component_menu {
        list-style: none;
    }

    &--isDivided {
        .List__item + .List__item {
            border-top: 1px solid var(--color-neutral-25);
        }
    }
}
