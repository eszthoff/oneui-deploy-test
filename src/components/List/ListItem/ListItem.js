import React from 'react';
import PropTypes from 'prop-types';
import bem from 'bem';
import Text from '../../Text';
import styles from './ListItem.scss';

const { block, elem } = bem({
    name: 'ListItem',
    classnames: styles,
    propsToMods: ['isDivided', 'isDividedList', 'isSelected', 'onClick']
});

const ListItem = props => {
    const { children, isDivided, isDividedList, isSelected, onClick, ...rest } = props;
    const customBlockMod = { clickable: typeof onClick === 'function' };

    return (
        <li {...rest} {...block(props, customBlockMod)}>
            <div onClick={onClick} role="presentation" {...elem('container', props)}>
                {React.Children.map(
                    children,
                    child => (typeof child === 'string' ? <Text>{child}</Text> : child)
                )}
            </div>
        </li>
    );
};

ListItem.displayName = 'ListItem';

ListItem.propTypes = {
    /** List Items */
    children: PropTypes.node,
    /** Formats this item as selected  */
    isSelected: PropTypes.bool,
    /** Adds a dividing line below the items. Do not use on divided lists */
    isDivided: PropTypes.bool,
    /** For internal use: adds a dividing line below all items but the last. isDivided set to true will overwrite this param */
    isDividedList: PropTypes.bool,
    /** A function to be called if the item is clicked */
    onClick: PropTypes.func
};

ListItem.defaultProps = {
    children: null,
    isSelected: false,
    isDivided: false,
    isDividedList: false,
    onClick: null
};

export default ListItem;
