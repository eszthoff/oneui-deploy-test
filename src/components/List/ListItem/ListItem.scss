.ListItem {
    transition: background-color var(--transition-duration);

    &__container {
        display: flex;
        align-items: baseline;
        padding: var(--spacing-3x);
    }

    &:hover {
        background-color: var(--color-hover-background);
    }

    &--isSelected {
        background-color: var(--color-highlight-background);
    }

    &--clickable {
        cursor: pointer;
    }

    &--disabled:hover {
        background-color: var(--color-background);
    }
}
