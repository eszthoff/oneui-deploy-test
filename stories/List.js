import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, select, withKnobs } from '@storybook/addon-knobs';
import { List, ListItem, ListActions, JobResult, Text, Button, Checkbox } from '@textkernel/oneui';

storiesOf('List', module)
    .addDecorator(withKnobs)
    .add('List', () => (
        <List
            Component={select('List component', ['ul', 'ol', 'menu'], 'ul')}
            isDivided={boolean('Add dividers between items', true)}
        >
            <ListItem>
                First item passed as string{' '}
                <ListActions>
                    <Button context="link">Action</Button>
                </ListActions>
            </ListItem>
            <ListItem>
                <Checkbox id="item 2" />
                <div>
                    <Text>Item passed as Text component</Text>
                    <Text>It has multiple lines</Text>
                </div>
            </ListItem>
            <ListItem isSelected={boolean('Select item', true)}>
                <Text inline>Item selectable via props</Text>
            </ListItem>
            <ListItem
                onClick={e => {
                    e.preventDefault();
                    alert('Item was selected'); // eslint-disable-line no-alert, no-undef
                }}
            >
                <Checkbox id="item 4" />
                Clickable item
            </ListItem>
        </List>
    ))
    .add('List showing job results', () => {
        const onClickJob = e => {
            e.preventDefault();
            alert('Item was selected'); // eslint-disable-line no-alert, no-undef
        };
        const onClickAction = e => {
            e.preventDefault();
            e.stopPropagation();
            alert('Action button was clicked'); // eslint-disable-line no-alert, no-undef
        };

        return (
            <div style={{ width: '500px' }}>
                <List isDivided>
                    <ListItem onClick={onClickJob}>
                        <JobResult
                            title="My first job"
                            location=" - Melbourne"
                            organization="Awsome inc."
                            details="It was posted here, yesterday"
                        />
                        <ListActions>
                            <Button context="link" onClick={onClickAction}>
                                Action
                            </Button>
                        </ListActions>
                    </ListItem>
                    <ListItem onClick={onClickJob}>
                        <JobResult
                            title="Job with a very, very, very long title to be truncated by css"
                            location=" - Melbourne"
                            organization="Awsome inc."
                            details="It was posted there, yesterday"
                        />
                        <ListActions>
                            <Button context="link" onClick={onClickAction}>
                                Action
                            </Button>
                        </ListActions>
                    </ListItem>
                    <ListItem onClick={onClickJob}>
                        <JobResult title="An other job with a very, very, very long title to be truncated by css if it doesn't fit" />
                        <ListActions>
                            <Button context="link" onClick={onClickAction}>
                                Action
                            </Button>
                        </ListActions>
                    </ListItem>
                </List>
            </div>
        );
    });
