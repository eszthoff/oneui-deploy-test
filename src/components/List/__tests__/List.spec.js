import React from 'react';
import toJson from 'enzyme-to-json';
import ListItem from '../ListItem';
import List from '../List';

describe('List component', () => {
    it('should render List correctly', () => {
        const wrapper = mount(
            <List>
                <ListItem>Item 1</ListItem>
                <ListItem>Item 2</ListItem>
            </List>
        );

        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('ul')).toHaveLength(1);
    });
    it('should render with ordered list correctly', () => {
        const wrapper = mount(
            <List Component="ol">
                <ListItem>Item 1</ListItem>
                <ListItem>Item 2</ListItem>
            </List>
        );

        expect(wrapper.find('ol')).toHaveLength(1);
    });
});
