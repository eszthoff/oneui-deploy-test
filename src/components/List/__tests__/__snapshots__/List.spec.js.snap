// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should render List correctly 1`] = `
<List
  Component="ul"
  isDivided={false}
>
  <ul
    className="List List--Component List--Component_ul"
  >
    <ListItem
      className="List__item List__item--Component List__item--Component_ul"
      isSelected={false}
      key=".0"
      onClick={null}
    >
      <li
        className="ListItem List__item List__item--Component List__item--Component_ul"
      >
        <div
          className="ListItem__container"
          onClick={null}
          role="presentation"
        >
          <Text
            inline={true}
            key=".0"
            muted={false}
            size="normal"
          >
            <span
              className="Text Text--size Text--size_normal"
              muted={false}
            >
              Item 1
            </span>
          </Text>
        </div>
      </li>
    </ListItem>
    <ListItem
      className="List__item List__item--Component List__item--Component_ul"
      isSelected={false}
      key=".1"
      onClick={null}
    >
      <li
        className="ListItem List__item List__item--Component List__item--Component_ul"
      >
        <div
          className="ListItem__container"
          onClick={null}
          role="presentation"
        >
          <Text
            inline={true}
            key=".0"
            muted={false}
            size="normal"
          >
            <span
              className="Text Text--size Text--size_normal"
              muted={false}
            >
              Item 2
            </span>
          </Text>
        </div>
      </li>
    </ListItem>
  </ul>
</List>
`;
